presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java

```diff
@@ -74,18 +75,25 @@
     private final List<Type> types;
     private final LongArrayList valueAddresses;
     private final ObjectArrayList<Block>[] channels;
+    private final boolean eagerCompact;
```

Comment:
I think we might need something more elaborate than true/false flag. My understanding of currently implemented approach was that eager compacting is redundant overhead when there is not so much data to be stored in `PagesIndex`. Thus, optimal solution could be to compact when certain threshold is reached.

Also, compacting may incur constant overhead (currently it's strictly incremental, I know), so you could want to avoid calling compact on every page. For example, compacting could merge small blocks into bigger ones for better memory utilization. Again, deferring next compact call until certain amount of data was received since the last one could be beneficial.

Reply:
I totally agree with you! We had a design discussion with @dain  and @haozhun , and here is a quick summary:

- Today's lazy compact doesn't work as expect in some corner cases. For example, a query is close to memory limit and it has both `HashBuilderOperator` (supports compact) and `GroupByOperator` (doesn't support compact). Whether the query will do compact or fail depends on chance: whether the memory reservation failure is triggered by  `HashBuilderOperator` or `GroupByOperator`

- Thus, we want to do compact deterministically. Our original decision is to have a config threshold so compact will eagerly happen when memory used on node exceeds some threshold . However, we find this very difficult to implement without a major change to today's memory management model. (It needs to go 5 ways up, `OperatorContext` -> `DriverContext` -> `PipelineContext` -> `TaskContext` -> `QueryContext` to get the current memory reservation. So we decide to start with this simple true/false.

- Eager compact will result in more CPU usage due to the copy cost. We should avoid unnecessary compact. For example, for block already in compact form, we can skip it (https://github.com/prestodb/presto/pull/8458). A future work would be only compact blocks that waste, e.g. >10% memory.

- Compact consecutive pages is really a great idea we didn't think! Any comments on this @dain , @haozhun ?

Reply:
@wenleix makes sense. I didn't intend to make `PagesIndex` aware of node-level memory state. Indeed, it would make _shall-i-compact?_ decision more educated rather than size-based heuristic, but more coupling. Thus i thought about a heuristic. Your idea to compact "blocks that waste, e.g. >10% memory" sounds also interesting.

I was also thinking about short vs long queries. A short query is less likely to run out of memory (it wouldn't be that short) and impact of CPU cycles spent on compacting might be more visible. For long query (like sizeable join), eager compacting is obvious thing to do, because you will keep this memory for long time (AFAIU the blocks in `PagesIndex` remain as-is even after `LookupSource` is built, so please correct me if I'm wrong on this). Short vs long queries can be distinguished at planning time with the use of statistics.

> Compact consecutive pages is really a great idea

I actually stole it from @arhimondr. He introduced it as a way to reduce per-page processing overhead (so nothing to do with memory savings per se). I am only speculating on potential memory savings it may imply. Per-block overhead on instance size, RLE data. This would be big deal if blocks/pages had only very few rows.

In fact, if #8691 lands, `PagesIndex` should not encounter pages ever, so maybe we don't need any new logic here